### PR TITLE
Update follow feed to be infinite

### PR DESCRIPTION
**Pull request by piratekev**
This PR updates the code so `GetPostsForFollowFeedForPublicKey` will always return the number of posts requested, unless there are no posts remaining. 

`GetPostsForFollowFeedForPublicKey` works by looping through the list of followed public keys, calling `DBGetAllPostsAndCommentsForPublicKeyOrderedByTimestamp` for a `publicKey`, `minTimestampNanos` and `maxTimestampNanos`, then sorting the returned posts into a list and truncating at index `numToFetch`. If there are fewer posts than requested or no posts in time requested days, then it will return a smaller or empty list. 

Previously, `maxTimestampNanos` was always set to the current time and `minTimestampNanos` was always 2 days ago. `maxTimestampNanos` still defaults to the current time, but is updated to the timestamp of the last pagination key `startAfterPostHash`. `minTimestampNanos` is updated to default to 4 days instead of 2, but this number increases exponentially until the range is infinite if not enough posts are found.

### CODE REVIEW

File: cmd/run.go [hunk 1]
```diff
@@ -163,6 +163,10 @@ func init() {
 	// Tag transaction with node source
 	runCmd.PersistentFlags().Uint64("node-source", 0, "Node ID to tag transaction with. Maps to ../core/lib/nodes.go")
 
+	// Follow Feed Length
+    runCmd.PersistentFlags().Bool("follow-feed-infinite", false, "Return the entire follow feed")
```
Comment (lgbello) [inline cmd/run.go]: Don't we want this to be true?

Comment (piratekev) [reply cmd/run.go]: This is the default value. We'll set it to true in our helm config.

File: cmd/run.go [hunk 2]
```diff
@@ -163,6 +163,10 @@ func init() {
 	// Tag transaction with node source
 	runCmd.PersistentFlags().Uint64("node-source", 0, "Node ID to tag transaction with. Maps to ../core/lib/nodes.go")
 
+	// Follow Feed Length
+	runCmd.PersistentFlags().Bool("follow-feed-infinite", false, "Return the entire follow feed")
```
Comment (jlmart88) [inline cmd/run.go]: does this override the follow-feed-length? if so, then we should just use a single flag, and have either -1 or 0 represent "infinity"

File: routes/exchange.go [hunk 3]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
 		return nil, errors.Wrapf(err, "GetPostsForFollowFeedForPublicKey: Problem filtering out restricted public keys: ")
 	}
 
-	minTimestampNanos := uint64(time.Now().UTC().AddDate(0, 0, -2).UnixNano()) // two days ago
-	// For each of these pub keys, get their posts, and load them into the view too
-	for _, followedPubKey := range filteredPubKeysMap {
+	var postEntriesForFollowFeed []*lib.PostEntry
+	n := 1
+	for n < 6 {
```
Comment (jlmart88) [inline routes/exchange.go]: `for n := 1 ; n < 6; n++; {`
Also can you find a better variable name for readability

File: routes/exchange.go [hunk 4]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
...
+		postEntriesForFollowFeed = nil
+		daysToSubtract := int(math.Pow(4, float64(n)))
 
-		_, dbPostAndCommentHashes, _, err := lib.DBGetAllPostsAndCommentsForPublicKeyOrderedByTimestamp(
-			bav.Handle, fes.blockchain.Snapshot(), followedPubKey, false /*fetchEntries*/, minTimestampNanos, 0, /*maxTimestampNanos*/
-		)
-		if err != nil {
-			return nil, errors.Wrapf(err, "GetPostsForFollowFeedForPublicKey: Problem fetching PostEntry's from db: ")
+		minTimestampNanos := uint64(time.Now().UTC().AddDate(0, 0, -daysToSubtract).UnixNano()) // two days ago
```
Comment (jlmart88) [inline routes/exchange.go]: comment is wrong on this line

File: routes/exchange.go [hunk 5]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
...
+			maxTimestampNanos = bav.GetPostEntryForPostHash(startAfterPostHash).TimestampNanos
+			minTimestampNanos = uint64(time.Unix(0, int64(maxTimestampNanos)).AddDate(0, 0, -daysToSubtract).UnixNano()) // casting timestamp uint64 to int64 won't crash until 2262
 		}
 
-		// Iterate through the entries found in the db and force the view to load them.
-		// This fills in any gaps in the view so that, after this, the view should contain
-		// the union of what it had before plus what was in the db.
-		for _, dbPostOrCommentHash := range dbPostAndCommentHashes {
-			bav.GetPostEntryForPostHash(dbPostOrCommentHash)
+		if n > 4 {
```
Comment (jlmart88) [inline routes/exchange.go]: this seems arbitrary; can you leave a comment

File: routes/exchange.go [hunk 6]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
...
 	}
 
-	minTimestampNanos := uint64(time.Now().UTC().AddDate(0, 0, -2).UnixNano()) // two days ago
-	// For each of these pub keys, get their posts, and load them into the view too
-	for _, followedPubKey := range filteredPubKeysMap {
+	var postEntriesForFollowFeed []*lib.PostEntry
+	n := 1
+	for n < 6 {
+		postEntriesForFollowFeed = nil
+		daysToSubtract := int(math.Pow(4, float64(n)))
```
Comment (jlmart88) [inline routes/exchange.go]: why is this 4? make sure to leave comments for magic numbers, or even better: pull them out to a variable for readability

File: routes/exchange.go [hunk 7]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
...
+			}
+
+			if _, isFollowedByUser := followedPubKeysMap[lib.MakePkMapKey(postEntry.PosterPublicKey)]; isFollowedByUser {
+				postEntriesForFollowFeed = append(postEntriesForFollowFeed, postEntry)
+			}
 		}
 
-		if _, isFollowedByUser := followedPubKeysMap[lib.MakePkMapKey(postEntry.PosterPublicKey)]; isFollowedByUser {
-			postEntriesForFollowFeed = append(postEntriesForFollowFeed, postEntry)
+		if len(postEntriesForFollowFeed) > numToFetch {
```
Comment (jlmart88) [inline routes/exchange.go]: I think this should be `>=`

File: routes/exchange.go [hunk 8]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
 		return nil, errors.Wrapf(err, "GetPostsForFollowFeedForPublicKey: Problem filtering out restricted public keys: ")
 	}
 
-	minTimestampNanos := uint64(time.Now().UTC().AddDate(0, 0, -2).UnixNano()) // two days ago
-	// For each of these pub keys, get their posts, and load them into the view too
-	for _, followedPubKey := range filteredPubKeysMap {
+	var postEntriesForFollowFeed []*lib.PostEntry
+	n := 1
+	for n < 6 {
+		postEntriesForFollowFeed = nil
```
Comment (jlmart88) [inline routes/exchange.go]: so we are resetting this on each loop, and then re-querying the entire date range, until we reach the `numToFetch` value? why can't we keep adding onto this array until we reach the desired amount?

File: routes/exchange.go [hunk 9]
```diff
@@ -1617,41 +1618,63 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
...
-	// Iterate over the view. Put all posts authored by people you follow into an array
-	var postEntriesForFollowFeed []*lib.PostEntry
-	for _, postEntry := range bav.PostHashToPostEntry {
-		// Ignore deleted or hidden posts and any comments.
-		if postEntry.IsDeleted() || (postEntry.IsHidden && skipHidden) || len(postEntry.ParentStakeID) != 0 {
-			continue
+		// For each of these pub keys, get their posts, and load them into the view too
+		for _, followedPubKey := range filteredPubKeysMap {
+
+			_, dbPostAndCommentHashes, _, err := lib.DBGetAllPostsAndCommentsForPublicKeyOrderedByTimestamp(
```
Comment (jlmart88) [inline routes/exchange.go:1648]: is this query not paginated? this seems like an easy way to run out of memory

**Review by jlmart88 — CHANGES_REQUESTED** [review]

File: routes/exchange.go (hunk 9 above)
Comment (piratekev) [reply routes/exchange.go:1648]: `minTimestampNanos` and `maxTimestampNanos` are the only options available to paginate this query. That's why I went with this approach to start with small intervals and ramp up until there are enough posts

File: routes/exchange.go [hunk 10]
```diff
@@ -1617,48 +1618,72 @@ func (fes *APIServer) GetPostsForFollowFeedForPublicKey(bav *lib.UtxoView, start
...
+	maxDaysToSubtractExponent := 5 // 4^5 = 1024 days
+	for daysToSubtractExponent := 1; daysToSubtractExponent < 6; daysToSubtractExponent++ {
+		daysToSubtract := int(math.Pow(float64(baseDaysToSubtract), float64(daysToSubtractExponent))) // baseDaysToSubtract ^ daysToSubtractExponent
+		postsFetched := len(postEntriesForFollowFeed)
+
+		minTimestampNanos := uint64(time.Now().UTC().AddDate(0, 0, -daysToSubtract).UnixNano()) // baseDaysToSubtract ^ daysToSubtractExponent days ago
+		maxTimestampNanos := uint64(0)
+
+		if postsFetched > 0 {
+			maxTimestampNanos = postEntriesForFollowFeed[postsFetched-1].TimestampNanos                                  // set to last fetched
```
Comment (jlmart88) [inline routes/exchange.go]: can you use the `minTimestampNanos` from the previous loop here? that would prevent needing to sort the entries on each loop, and would make this more accurate (i.e. not rescanning timestamps that were already scanned)

**Review by jlmart88 — APPROVED** [review]

**Pull request merged**